### PR TITLE
reimplementation of command line interface

### PR DESCRIPTION
**Pull request by gertjanvanzwieten**

### CODE REVIEW

File: nutils/util.py [hunk 1]
```diff
@@ -437,6 +438,44 @@ def binaryfile(path):
...
+    if not changed:
+        return f
+    sig = sig.replace(parameters=params)
+    def defaults_from_env(*args, **kwargs):
+        bound = sig.bind(*args, **kwargs)
+        bound.apply_defaults()
+        return f(*bound.args, **bound.kwargs)
+    defaults_from_env.__signature__ = sig
+    defaults_from_env.__wrapped__ = f
+    defaults_from_env.__doc__ = f.__doc__
```
Comment (joostvanzwieten) [inline nutils/util.py:477]: Why not `@functools.wraps(f)` decorate the wrapper? `@functools.wraps` copies more than `__wrapped__` and `__doc__`, for example: `__module__`, `__name__`, `__qualname__`.

File: nutils/util.py [hunk 2]
```diff
@@ -437,6 +438,44 @@ def binaryfile(path):
     raise TypeError('binaryfile requires a path-like or file-like argument')
 
 
+def defaults_from_env(f):
+    '''Decorator for changing function defaults based on environment.
+
+    This decorator searches the environment for variables matching the pattern
+    ``NUTILS_MYPARAM``, where ``myparam`` is a parameter of the decorated
+    function. Only parameters with type annotation and a default value are
+    considered, and the string value is deserialized using stringly. In case
```
Comment (joostvanzwieten) [inline nutils/util.py:449]: Suggestion: Add a reference to the PyPI page of Stringly.

Comment (joostvanzwieten) [inline nutils/util.py:449]: Stringly capitalized?

File: nutils/util.py [hunk 3]
```diff
@@ -512,6 +514,23 @@ def set_current(*args, **kwargs):
...
+    fmt = '%Y-%m-%d %H:%M:%S'
+    treelog.info(f'start {time.strftime(fmt,time.localtime(t0))}')
+    try:
+        yield
+    finally:
+        te = time.time()
+        m, s = divmod(int(te - t0), 60)
+        if m > 60:
+            m = '{}:{:02d}'.format(*divmod(m, 60))
+        treelog.info(f'finish {time.strftime(fmt,time.localtime(t0))}, elapsed {m}:{s:02d}')
```
Comment (joostvanzwieten) [inline nutils/util.py:531]: Alternatively you can use `datetime`. Advantage: you can write the time format of a `datetime` as format parameter:

```python
t0 = datetime.now()
treelog.info(f'start {t0:%Y-%m-%d %H:%M:%S}')
```

The difference between two `datetime` instances, `te - t0`, has a `total_seconds()` method.

File: nutils/cli.py [hunk 4]
```diff
@@ -1,397 +1,361 @@
 """
-The cli (command line interface) module provides the `cli.run` function that
+he cli (command line interface) module provides the `cli.run` function that
```
Comment (joostvanzwieten) [inline nutils/cli.py:2]: Typo.

File: nutils/cli.py [hunk 5]
```diff
@@ -1,397 +1,361 @@
...
+        params.append(param)
+        cnames.append(param.name)
 
+    def in_context(*args, **kwargs):
+        with context(**{name: kwargs.pop(name) for name in cnames if name in kwargs}):
+            f(*args, **kwargs)
+
+    in_context.__signature__ = sig.replace(parameters=params)
+    in_context.__wrapped__ = f
+    in_context.__doc__ = f.__doc__
```
Comment (joostvanzwieten) [inline nutils/cli.py:48]: Why not use `@functools.wraps(f)` here. See earlier comment.

File: nutils/util.py (hunk 1 above)
Comment (gertjanvanzwieten) [reply nutils/util.py:477]: It does not introduce the new signature though. Better to combine wraps with assigning `__signature__` only?

Comment (joostvanzwieten) [reply nutils/util.py:477]: That's better IMO. If I'm not mistaken, Sphinx will ignore functions in a module if the `__module__` attribute doesn't match the global `__name__`.

Comment (gertjanvanzwieten) [reply nutils/util.py]: fixed

File: nutils/util.py (hunk 2 above)
Comment (gertjanvanzwieten) [reply nutils/util.py]: fixed

File: nutils/cli.py (hunk 4 above)
Comment (gertjanvanzwieten) [reply nutils/cli.py]: fixed

File: nutils/cli.py (hunk 5 above)
Comment (gertjanvanzwieten) [reply nutils/cli.py]: fixed

**Review by joostvanzwieten — APPROVED** [review]

**Pull request merged**